SerialPrograms/Source/CommonFramework/AudioPipeline/AudioThreadController.cpp

```diff
@@ -0,0 +1,600 @@
+                    break;
+#endif
+                }
+            });
+        }
+    }
+
+    connect(m_audioIODevice, &AudioIODevice::fftInputReady, this, &AudioWorker::fftInputReady);
+}
+
+AudioWorker::~AudioWorker(){
+    if (m_audioIODevice){
+        // Close the connection between m_audioIODevice and 
+        // m_audioSink.
+        m_audioIODevice->setAudioSinkDevice(nullptr);
+    }
+
+    if (m_audioSink){
+        m_audioSink->stop();
+        delete m_audioSink;
+
+        m_audioSink = nullptr;
+    }
+
+    if (m_audioSource){
+        m_audioSource->stop();
+        delete m_audioSource;
+        m_audioSource = nullptr;
+    }
+
+    if (m_audioIODevice){
+        delete m_audioIODevice;
+        m_audioIODevice = nullptr;
+    }
+}
+
+FFTWorker::FFTWorker(int fftLengthPowerOfTwo) : m_fftLengthPowerOfTwo(fftLengthPowerOfTwo){
+    const int numFFTSamples = 1 << m_fftLengthPowerOfTwo;
+    m_fftInputBuffer = AlignedVector<float>(numFFTSamples);
+    m_fftOutputBuffer = AlignedVector<float>(numFFTSamples);
+
+    m_outputVector.resize(numFFTSamples / 2);
+}
+
+FFTWorker::~FFTWorker(){}
+
+void FFTWorker::computeFFT(const QVector<float>& fftInput){
+    // float sum = 0.0;
+    // for(auto v : rawAudioSamples) sum += v;
+    // std::cout << "FFTWorker::computeFFT, receive fft input, sum " << sum << std::endl;;
+
+    // We need to copy data from one buffer to another because m_fftInputBuffer is 
+    // memory aligned. This is required for the FFT function.
+    memcpy(m_fftInputBuffer.data(), fftInput.data(), sizeof(float) * m_fftInputBuffer.size());
+
+    // raw audio samples are in range [-1.0, 1.0]. Need to convert it to [0.0, 1.0]
+    // for(size_t i = 0; i < m_fftInputBuffer.size(); i++){
+    //     m_fftInputBuffer[i] = (rawAudioSamples[i] + 1.0) / 2.0;
+    // }
+
+    auto startTime = std::chrono::system_clock::now();
+
+#ifdef USE_FFTREAL
+    m_fft.calculateFFT(m_fftOutputBuffer.data(), m_fftInputBuffer.data());
+
+    const auto& audioFormat = m_widget->m_audioFormat;
+    for(int i = 2; i <= NUM_FFT_SAMPLES/2; i++){
+        float real = m_fftOutputBuffer[i];
+        float imag = 0.0;
+        if (i>0 && i<NUM_FFT_SAMPLES/2) {
+            imag = m_fftOutputBuffer[NUM_FFT_SAMPLES/2 + i];
+        }
+        float mag = std::sqrt(real * real + imag * imag);
+        // std::cout << mag << " ";
+        m_fftOutputBuffer[i-2] = mag;
+    }
+#else
+    Kernels::AbsFFT::fft_abs(m_fftLengthPowerOfTwo, m_fftOutputBuffer.data(), m_fftInputBuffer.data());
+#endif
+
+    auto endTime = std::chrono::system_clock::now();
+    std::chrono::microseconds dur = std::chrono::duration_cast<std::chrono::microseconds>(endTime - startTime);
+    // std::cout << "T" << QThread::currentThread() << " FFTWorker::computeFFT() finished, time " dur.count() <<
+    //     " ms" << std::endl;
+
+    memcpy(m_outputVector.data(), m_fftOutputBuffer.data(), sizeof(float) * m_outputVector.size());
+    emit FFTFinished(m_outputVector);

```

Comment:
Is this safe to pass by reference? I'm not sure how signals handle references. But this is running on the FFT thread right? And the signal is being emitted to the UI thread.

Since you're passing by reference, `m_outputVector` needs to live long enough for the receiving thread to use it. If the FFT thread yeets it before then, we're screwed.

Normally, I pass all signals by value unless I know the thing that's being passed will outlive the signal receiver.

Reply:
I looked it up before coding this. I found this post which found out Qt makes internal copies of the parameters when it is passed across different threads: https://stackoverflow.com/questions/8455887/stack-object-qt-signal-and-parameter-as-reference
If this post is true, we can safely pass in const references to remove one copy of the data.